Add coverage for `power.py` test

bc1cindy opened the pull request:
Cover missing battery calculation for generic devices and USB detection. Using unit tests instead of functional pattern as PowerManager is a utility class without UI - no button sequences or user interaction to test.

### What is this PR for?



### Changes made to:
- [ ] Code
- [x] Tests
- [ ] Docs
- [ ] CHANGELOG


### Did you build the code and tested on device?
- [ ] Yes


### What is the purpose of this pull request?
- [ ] Bug fix
- [ ] New feature
- [ ] Docs update
- [ ] Other

qlrd reviewed (changes requested):
I think the assertions are sufficient, but mixed too much the tests and the mocks are improper. Also, there's only a device being tested

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,60 @@ def test_reboot_without_pmu(mocker, m5stickv):
     krux.power.sys.exit.assert_called()
+
+
+def test_battery_charge_remaining_with_charging(mocker, multiple_devices):
+    from krux.power import PowerManager
+    import board
+
+    manager = PowerManager()
+
+    if manager.pmu is not None:
+        manager.pmu.get_battery_voltage.return_value = 4000
+        manager.pmu.charging.return_value = True
+
+        charge = manager.battery_charge_remaining()
+
+        assert charge <= 0.9
```
instead use `multiple_devices` fixture, is better to test each device that has `pmu` object. Also, is good to `spy` some internal calls and assert that they are called. Made a "PR in your PR"

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,60 @@ def test_reboot_without_pmu(mocker, m5stickv):
+        (4200, 1.0),
+    ],
+)
+def test_battery_charge_remaining_formula(mocker, multiple_devices, voltage, expected):
+    from krux.power import PowerManager
+    import board
+
+    manager = PowerManager()
+
+    if manager.pmu is not None:
+        manager.pmu.get_battery_voltage.return_value = voltage
+        manager.pmu.charging.return_value = False
+
+        charge = manager.battery_charge_remaining()
+
+        if board.config["type"] in ("amigo", "m5stickv", "cube"):
+            assert 0.0 <= charge <= 1.0
+        else:
+            assert charge == expected
```
same, but calling a "newer device"

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,60 @@ def test_reboot_without_pmu(mocker, m5stickv):
+            assert charge == expected
+
+
+def test_usb_connected(mocker, multiple_devices):
+    from krux.power import PowerManager
+
+    manager = PowerManager()
+
+    if manager.pmu is not None:
+        manager.pmu.usb_connected.return_value = True
+        assert manager.usb_connected() is True
+        manager.pmu.usb_connected.assert_called_once()
+
+        manager.pmu.usb_connected.reset_mock()
+        manager.pmu.usb_connected.return_value = False
+        assert manager.usb_connected() is False
+        manager.pmu.usb_connected.assert_called_once()
```
separated it in "pairing tests"

qlrd commented on src/krux/power.py:
```diff
@@ -59,11 +59,25 @@ def has_battery(self):
+        # Start to calculate charge
         mv = int(self.pmu.get_battery_voltage())
+
+        # For amigo and m5stickv he had done testing
```
```suggestion
        # For amigo and m5stickv we had done testing
```

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,188 @@ def test_reboot_without_pmu(mocker, m5stickv):
+# This test is for potentially new devices
+# that are not implemented yet in the firmware
+# and do not have the necessary linear approximation
+# # to calculate the battery charge remaining.
```
extra `#` here

```suggestion
# to calculate the battery charge remaining.
```

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,188 @@ def test_reboot_without_pmu(mocker, m5stickv):
+# and do not have the necessary linear approximation
+# # to calculate the battery charge remaining.
+# It will use the MAX_BATTERY_MV and MIN_BATTERY_MV
+# # to calculate the charge remaining. and work
```
here too

odudex commented on src/krux/power.py:
```diff
@@ -59,11 +59,25 @@ def has_battery(self):
+        # unbound value when calculating charge
+        charge = 1.0
+
+        # Start to calculate charge
```
IMO these comments could be removed

bc1cindy replied:
removed @odudex

bc1cindy commented:
removed @odudex

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,188 @@ def test_reboot_without_pmu(mocker, m5stickv):
+# and do not have the necessary linear approximation
+# to calculate the battery charge remaining.
+# It will use the MAX_BATTERY_MV and MIN_BATTERY_MV
+# to calculate the charge remaining. and work
```
```suggestion
# to calculate the charge remaining and work
```

bc1cindy commented:
Fixed both issues @odudex 

1. Removed duplicate function call - now calling battery_charge_remaining() only once
3. Changed assertions to use exact values with pytest.approx() tolerance instead of <=

Also switched from mocker.spy to mocker.patch.object to properly mock return values.

Thank you.

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,155 @@ def test_reboot_without_pmu(mocker, m5stickv):
+    # (voltage, expected_charge)
+    cases = [
+        (4200, 1.0),
+        (4000, 1.0),
```
IMO, the second argument should be less than 1.0

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,155 @@ def test_reboot_without_pmu(mocker, m5stickv):
+    ]
+
+    for voltage, expected in cases:
+        mocker.patch.object(manager.pmu, "get_battery_voltage", return_value=voltage)
+        mocker.patch.object(manager.pmu, "charging", return_value=False)
```
What's is the reason to swap the `mock.spy` with `mocker.patch`?

IMO mocking the return calls, in this context, do not really test if the internal functions (`get_battery_voltage` and `charging`) was called correctly. 

This force a result to be a "would like" thing not what "should be", maybe isn't better to mock the `battery` itself like before? 

Also, this not checking if these internal methods are being called correctly in their argument calls (`assert_called_once_with`).

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,155 @@ def test_reboot_without_pmu(mocker, m5stickv):
+    ]
+
+    for voltage, expected in cases:
+        mocker.patch.object(manager.pmu, "get_battery_voltage", return_value=voltage)
+        mocker.patch.object(manager.pmu, "charging", return_value=False)
```
Same here

qlrd commented on tests/test_power.py:
```diff
@@ -146,3 +146,155 @@ def test_reboot_without_pmu(mocker, m5stickv):
+    ]
+
+    for voltage, expected in cases:
+        mocker.patch.object(manager.pmu, "get_battery_voltage", return_value=voltage)
+        mocker.patch.object(manager.pmu, "charging", return_value=False)
```
Same here

bc1cindy commented:
You're right @qlrd! I've updated the tests to:

1. Use MagicMock directly
2. Add spy tracking, to verify method calls
3. Include assert_called_once(), for proper call verification

New approach:
manager.pmu.get_battery_voltage = mocker.MagicMock(return_value=voltage)
voltage_spy = mocker.spy(manager.pmu, "get_battery_voltage")
# ... test ...
voltage_spy.assert_called_once()

This maintains consistency while properly testing the internal method calls. Thanks for the feedback!
